Improved metatag usage

dominiclord opened the pull request:
- Alternate language listing
- Allow hooking into final metatag methods
- Configurable reCaptcha lang

dominiclord committed:
Renderable alternate translations (notably in cms/metatag/document.ms)

dominiclord committed:
Safer validation before using url() method on contextObject

dominiclord committed:
Set lang on root URL in alternate translations

dominiclord committed:
New hooks for metatag related methods allow for local application config overrides

dominiclord committed:
i18n of recaptcha script

dominiclord committed:
l10n properties not looping when used as structure interface

mcaskill commented on src/Cms/AbstractWebTemplate.php:
```diff
@@ -289,4 +360,93 @@ public function twitterCardImage()
+
+                $this->localesManager->setCurrentLocale($lang);
+
+                $data = [
```
Let's move the array format to a separate method for easy customization:
`protected function formatAlternateTranslation()`

mcaskill commented on src/Cms/AbstractWebTemplate.php:
```diff
@@ -289,4 +360,93 @@ public function twitterCardImage()
+
+
+
+    // Polylingual
```
I propose we move these methods to a `LocaleAwareTrait` except for `AbstractWebTemplate::currentLanguage()` and `AbstractWebTemplate::currentLocale()`.

mcaskill commented on src/Cms/AbstractWebTemplate.php:
```diff
@@ -153,14 +175,31 @@ public function metaDescription()
+        }
+
+        if (!$desc) {
+            $desc = $this->metaDescriptionFromConfig();
```
Is there a reason the fallback methods are called `meta*FromConfig()`?
I think ``fallbackMeta*()` would be better.

dominiclord committed:
Defining more agnostic method names for fallback hooks

dominiclord committed:
Created a LocaleAwareTrait, abstracting access to available languages/translations

dominiclord commented:
Not quite sure how to implement your `formatAlternateTranslation` suggestion. Seems like a good idea though.

mcaskill committed:
Various improvements to LocaleAwareTrait

Changes:
- Decoupled alternate translation structure from `alternateTranslations()` to `formatAlternateTranslation()`
- Added abstract methods
- Fixed PHPCS issues

mcaskill committed:
Minor improvement to `LocaleAwareTrait::hasAlternateTranslations()`

mcaskill commented on src/Cms/LocaleAwareTrait.php:
```diff
@@ -0,0 +1,189 @@
+     */
+    protected function formatAlternateTranslation($context, $lang)
+    {
+        $isModel    = ($context instanceof ModelInterface);
```
Is there ever an occasion where we wouldn't be processing a `ModelInterface`?

I defined `ContextualTemplate::contextObject()` mixin to return a model.

dominiclord committed:
Tweaking Generator usage for building off alternate translations
